openssl_* module_utils/crypto.py: add full list of OIDs known to current OpenSSL

felixfontein opened the pull request:
##### SUMMARY
Adds a full list of OIDs known to OpenSSL. Created from https://github.com/openssl/openssl/blob/master/crypto/objects/objects.txt by https://gist.github.com/felixfontein/376748017ad65ead093d56a45a5bf376

This is a large piece over 1050 lines, but then at least we can serve the same names for all items listed independent of the concrete OpenSSL library used by the module.

I hope I've attributed this part correctly. While the numbers and names in this collection are kind of public, the collection itself should be copyrightable.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/crypto.py

felixfontein committed:
Add full list of OIDs known to current OpenSSL.

felixfontein committed:
Remove hardcoded OIDs.

MarkusTeufelberger commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -341,50 +345,1128 @@ def remove(self, module):
+# #####################################################################################
+# #####################################################################################
+# This has been extracted from the OpenSSL project's objects.txt:
+#     https://github.com/openssl/openssl/blob/master/crypto/objects/objects.txt
```
Maybe refer to the actual commit here?

felixfontein replied:
Good idea! Did that now.

felixfontein committed:
UID -> x500UniqueIdentifier

felixfontein committed:
Reference actual version used.

MarkusTeufelberger commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -341,50 +345,1128 @@ def remove(self, module):
+    '0.9.2342.19200300': ('ucl', ),
+    '0.9.2342.19200300.100': ('pilot', ),
+    '0.9.2342.19200300.100.1': ('pilotAttributeType', ),
+    '0.9.2342.19200300.100.1.1': ('userId', 'UID'),
```
uid number 1

and...

MarkusTeufelberger commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -341,50 +345,1128 @@ def remove(self, module):
+    '0.9.2342.19200300.100.1.41': ('mobileTelephoneNumber', ),
+    '0.9.2342.19200300.100.1.42': ('pagerTelephoneNumber', ),
+    '0.9.2342.19200300.100.1.43': ('friendlyCountryName', ),
+    '0.9.2342.19200300.100.1.44': ('uniqueIdentifier', 'uid'),
```
uid number 2. Which one will end up mapping to the OID?

felixfontein replied:
Hmmm. That's unfortunate. I was hoping I can ignore case, to work around the problem that OpenSSL uses `userId`, while cryptography uses `userID` (or was it vice versa?). Maybe we should remove the `lower()` everywhere and manually add the alias `userID` for `userId`. What do you think?

MarkusTeufelberger replied:
Removing `lower()` (and maybe adding a few mappings by hand) seems like the safer/cleaner option... `uid != UID` and `mail != Mail` is already bad enough in OpenSSL's OID mapping...

MarkusTeufelberger commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -341,50 +345,1128 @@ def remove(self, module):
-    for name in names[1:]:
-        _NORMALIZE_NAMES[name] = names[0]
+    for name in names:
+        _NORMALIZE_NAMES[name.lower()] = names[0]
```
Maybe something along these lines might be useful?

```suggestion
        assert name.lower() not in _NORMALIZE_NAMES or _NORMALIZE_NAMES[name.lower()] == names[0]
        _NORMALIZE_NAMES[name.lower()] = names[0]
```

MarkusTeufelberger replied:
Ah, those just map the shorter names to the longer ones...

felixfontein replied:
Usually yes. If there aren't collisions, such as `uid` and `UID`...

There's also another collision: both `1.3.6.1.7` (`'Mail'`) and `0.9.2342.19200300.100.1.3` (`'rfc822Mailbox', 'mail'`) have the lowercase name `mail`. Besides these two, there are no more collisions.

felixfontein replied:
Usually yes. If there aren't collisions, such as `uid` and `UID`...

There's also another collision: both `1.3.6.1.7` (`'Mail'`) and `0.9.2342.19200300.100.1.3` (`'rfc822Mailbox', 'mail'`) have the lowercase name `mail`. Besides these two, there are no more collisions.

felixfontein committed:
Don't normalize to lower-case.

MarkusTeufelberger commented on test/integration/targets/openssl_certificate_info/tasks/main.yml:
```diff
@@ -27,7 +27,7 @@
       title: Chief
       pseudonym: test
-      UID: asdf
+      x500UniqueIdentifier: asdf
```
That would have been a "UserID" (OID `0.9.2342.19200300.100.1.1`) before - I guess this change is intentional?

felixfontein replied:
It was, though I'm not sure whether it is still necessary (or whether that problem was caused by lowercasing). I'll add a commit to revert this, let's see what happens.

felixfontein replied:
Looks like it isn't necessary anymore. I've removed the same change from #54921.

felixfontein committed:
Change test back.

MarkusTeufelberger commented:
Looks great!

shipit

felixfontein commented:
@gundalow can you check whether the attribution of the data collected by OpenSSL is ok?

gundalow reviewed (changes requested):
Just marking as -1 till I get someone in Legal to approve this.

felixfontein committed:
Fix typo.

felixfontein commented:
The `new_module` / `new_plugin` is incorrect; I've just rebased to update to current `devel`, where a new module was recently merged (and which this PR now also modifies; ansible/ansibullbot#1167).

gundalow commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -15,11 +15,15 @@
 # Copyright (c) 2017 Fraser Tweedale (@frasertweedale)
 # For more details, search for the function _obj2txt().
+# -----------------------------------------------------------------------------
+# A clearly marked portion of this file is licensed under the Apache License 2.0
```
```suggestion
# A clearly marked portion of this file is extracted from a project that is licensed under the Apache License 2.0
```

felixfontein replied:
Done. I've added a line break to shorten this a bit (since the rest of the header is also not that long), I hope that's ok :)

gundalow commented on lib/ansible/module_utils/crypto.py:
```diff
@@ -355,42 +359,1133 @@ def remove(self, module):
+# This has been extracted from the OpenSSL project's objects.txt:
+#     https://github.com/openssl/openssl/blob/9537fe5757bb07761fa275d779bbd40bcf5530e4/crypto/objects/objects.txt
+# Extracted with https://gist.github.com/felixfontein/376748017ad65ead093d56a45a5bf376
+#
```
```suggestion
#
# In case the following data structure has any copyrightable content, note that it is licensed as follows:
```

felixfontein replied:
Done.

gundalow reviewed (changes requested):
Red Hat legal have suggested two changes.

felixfontein committed:
Apply changes suggested by RedHat legal.

felixfontein commented:
ready_for_review

felixfontein commented:
@MarkusTeufelberger can you take another look? The only things I changed are the copyright notice (according to RedHat legal team wishes) and I rebased and applied the same changes needed for `openssl_certificate_info` to `openssl_csr_info`.

MarkusTeufelberger commented:
:ship: it / :shipit: 

shipit

felixfontein commented:
> :shipit:

YMMD :D

gundalow commented:
Red Hat legal have given this a +1, so merging

felixfontein commented:
@MarkusTeufelberger thanks for reviewing!
@gundalow thanks for asking Red Hat legal, and thanks for merging!
